### PR TITLE
WIP[#5345]: RoleSelector field in ServiceMonitor (#5345)

### PR DESCRIPTION
**Pull request by yp969803**
## Description

If it fixes a bug or resolves a feature request, be sure to link to that issue. #5345 

Be able to define selectors in kubernetes_sd_config

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note

```

### CODE REVIEW

File: pkg/apis/monitoring/v1/servicemonitor_types.go [hunk 1]
```diff
@@ -75,7 +75,12 @@ type ServiceMonitorSpec struct {
 	Endpoints []Endpoint `json:"endpoints"`
 
 	// Label selector to select the Kubernetes `Endpoints` objects.
-	Selector metav1.LabelSelector `json:"selector"`
+	EndpointsSelector metav1.LabelSelector `json:"endpointsSelector"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/servicemonitor_types.go:78]: It's not possible to rename an existing field.

File: pkg/apis/monitoring/v1/servicemonitor_types.go [hunk 2]
```diff
@@ -162,6 +167,23 @@ type ServiceMonitorList struct {
 	Items []*ServiceMonitor `json:"items"`
 }
 
+// RoleSelector is a selector for Kubernetes resources.
+// +k8s:openapi-gen=true
+type RoleSelector struct {
+
+	// Role specifies the role of the Kubernetes resources to be selected.
+	// Possible values for Role are: "endpoints", "endpointslice", "service", "pod", "node", or "ingress".
+	Role string `json:"role" protobuf:"bytes,1,opt,name=role"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/servicemonitor_types.go:176]: no need for the protobuf tag.

File: pkg/apis/monitoring/v1/servicemonitor_types.go [hunk 3]
```diff
@@ -75,7 +75,12 @@ type ServiceMonitorSpec struct {
 	Endpoints []Endpoint `json:"endpoints"`
 
 	// Label selector to select the Kubernetes `Endpoints` objects.
-	Selector metav1.LabelSelector `json:"selector"`
+	EndpointsSelector metav1.LabelSelector `json:"endpointsSelector"`
+
+	// Selector is a list of RoleSelector.
+	// +optional
+	Selector []RoleSelector `json:"selector"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/servicemonitor_types.go:82]: I expect label/field selectors to be defined in the Endpoint field.

File: pkg/apis/monitoring/v1/servicemonitor_types.go [hunk 4]
```diff
@@ -162,6 +167,23 @@ type ServiceMonitorList struct {
 	Items []*ServiceMonitor `json:"items"`
 }
 
+// RoleSelector is a selector for Kubernetes resources.
+// +k8s:openapi-gen=true
+type RoleSelector struct {
+
+	// Role specifies the role of the Kubernetes resources to be selected.
+	// Possible values for Role are: "endpoints", "endpointslice", "service", "pod", "node", or "ingress".
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/servicemonitor_types.go:175]: We can add validation markers.

**Comment by ArthurSens** [comment]
Hey, I've read the linked issue but I couldn't understand why we would want different roles in ServiceMonitors that are not Endpoint or EndpointSlice. (Already being worked on in https://github.com/prometheus-operator/prometheus-operator/pull/6518)

I.e., if we're already using the `Endpoint` role (SM's default role), for example, why would we use a selector using a different role?

**Comment by yp969803** [comment]
@ArthurSens,  agree with you, actually, i have taken referenced from the prometheus doc https://prometheus.io/docs/prometheus/latest/configuration/configuration/#kubernetes_sd_config

**Comment by yp969803** [comment]
@ArthurSens @simonpasquier i have changed the things, can you look into the commit is I am going correct

**Comment by yp969803** [comment]
Sorry for delayed work busy with my college stuffs, closing this as @nicolastakashi is working on it!

**Pull request closed without merging**